Use one tool for reading items instead of a tool per collection

yoavf opened the pull request:
## The problem
Having a tool per collection can cause issues with MCP clients that have hard limits of number of tools available (i.e Cursor), with Directus installs with many collections.

<img width="906" alt="Screenshot 2025-04-08 at 12 51 44" src="https://github.com/user-attachments/assets/2135c0b7-b123-4bc5-a853-35da638e4bc6" />

Additionally, when there are many junction tables for many to many relationships, the combined server and tool name length can easily exceed 60 characters

<img width="860" alt="Screenshot 2025-04-08 at 12 52 39" src="https://github.com/user-attachments/assets/d8b79a83-dbf3-4e0c-8f5a-e5c149f39f99" />

## Suggested solution
As an alternative, we can have a single tool to read items from any collection.

Here's how the updated tool list looks like in Cursor:
<img width="942" alt="Screenshot 2025-04-08 at 12 54 46" src="https://github.com/user-attachments/assets/8ac44b53-14c1-4224-b83c-4e0e3dc4d39c" />

Commit:
Use one tool for reading items instead of a tool per collection

Having a single tool per collections can cause issues with MCP clients
that have hard limits of number of tools available (i.e Cursor),
with Directus installs with many collections.
Additionally, when there are a lot of junction tables for many_to_many
relationships, the collection names can grow too long and the
tool-per-collection name will exceed the max length for a tool name.

This update creates a single tool to read items that takes a collection
name as param

rijkvanzanten commented:
Ahh interesting! To make Claude work better I wanted to strong-type the input schema, but that in turn caused the input schema to be too large for a single tool. I see here that you dropped the strong-input type in favor of a weak typed input structure (eg just a string vs an enum of literals), is that going to be a problem? 🤔

yoavf commented:
Thanks for looking at this @rijkvanzanten . 

> I see here that you dropped the strong-input type in favor of a weak typed input structure (eg just a string vs an enum of literals), is that going to be a problem? 🤔

You're right, let me think about this for a bit :)

Commit:
read-items: implement runtime validation to address type safety

Adds runtime checks for collection, fields, and sort parameters within the
generic read-items tool handler. This provides type safety feedback
similar to the original per-collection tools but within the single
generic tool approach.

Refactors tool creation to accommodate passing the schema context to the handler.

yoavf commented:
@rijkvanzanten Let me know what you think about this improved implementation.

rijkvanzanten commented:
@yoavf Still not sure what the impact is of not having strictly typed input parameters, but having it work in cursor in the first place is more important than having it work slightly worse in Claude 🙂

rijkvanzanten commented:
Shipped as 0.1.0 🙂

yoavf commented:
Thank you!
